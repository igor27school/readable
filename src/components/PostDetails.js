import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { fetchPostFromServer, fetchPostCommentsFromServer } from '../actions'
import { compare } from '../utils/Helper'
import { POST_TYPE } from '../constants/ObjectTypes'
import Sorter from './Sorter'
import Voter from './Voter'
import Comment from './Comment'
import Deleter from './Deleter'

class PostDetails extends Component {
  componentDidMount() {
    const { posts } = this.props
    const { postId } = this.props.match.params
    if (postId && (!posts.byId || !(postId in posts.byId))) {
      this.props.fetchPostFromServer(postId).then(() => this.props.fetchPostCommentsFromServer(postId))
    } else if (postId && !('comments' in posts.byId[postId])) {
      this.props.fetchPostCommentsFromServer(postId)
    }
  }
  render() {
    const { sortOrder, posts, comments } = this.props
    const { postId } = this.props.match.params
    const post = posts.byId && posts.byId[postId]
    if (!post){
      return (
        <div>Post id is invalid or the post has been deleted: {postId}</div>
      )
    } else if (post.deleted) {
      return (
        <div>This post has been deleted: {postId}</div>
      )
    }
    const commentIds = post.comments ? post.comments.map(commentId => comments.byId[commentId]).filter(comment => !comment.deleted).sort(compare(sortOrder)).map(comment => comment.id) : []
    return (
      <div>
        <h2>
          <Deleter
            objectType={POST_TYPE}
            id={postId}
            history={this.props.history}
          />
          <Link to={`/edit/posts/${post.id}`}>Edit</Link>
          <span> {post.title}</span>
          <span> Score: {post.voteScore}</span>
          <Voter objectType={POST_TYPE} id={postId}/>
        </h2>
        <h3>{post.body}</h3>
        <h3>{post.author}</h3>
        <h3>{(new Date(post.timestamp)).toString()}</h3>
        {post.comments && post.comments.length > 0 && (
          <h3> Number of {post.comments.length === 1 ? 'comment' : 'comments'}: {post.comments.length}</h3>
        )}
        <Sorter/>
        {commentIds.map(commentId => (
          <Comment key={commentId} commentId={commentId}/>
        ))}
        <h4><Link to={`/create/${postId}`}>New Comment</Link></h4>
        <h4><Link to={`/categories/${post.category}`}>View {post.category} Category</Link></h4>
        <h4><Link to='/'>View All Categories</Link></h4>
      </div>
    )
  }
}

function mapStateToProps ({ sortOrder, posts, comments }) {
  return {sortOrder, posts, comments }
}

function mapDispatchToProps (dispatch) {
  return {
    fetchPostFromServer: (postId) => dispatch(fetchPostFromServer(postId)),
    fetchPostCommentsFromServer: (postId) => dispatch(fetchPostCommentsFromServer(postId)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(PostDetails)
