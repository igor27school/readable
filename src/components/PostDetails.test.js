import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { PostDetails } from './PostDetails'

jest.mock('../utils/ServerAPI')

describe('PostDetails tests', () => {
  const setup = () => {
    const props = {
      post: {
        id: 'testPostId',
        title: 'Test Post Title',
        body: 'testing PostDetails',
        author: 'Test Post Author',
        category: 'testCategory',
        timestamp: 1234,
        deleted: false,
        comments: ['testComment1', 'testComment2'],
        hasAllComments: true,
      },
      postId: 'testPostId',
      commentIds: ['testComment1', 'testComment2'],
      fetchPostFromServer: jest.fn(),
      fetchPostCommentsFromServer: jest.fn(),
      history: {},
    }
    const renderer = new ShallowRenderer()
    return {
      props,
      renderer
    }
  }

  it('PostDetails matches the snapshot', () => {
    const { renderer, props } = setup()
    renderer.render(<PostDetails {...props}/>)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
