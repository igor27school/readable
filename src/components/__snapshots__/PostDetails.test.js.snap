// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetails tests PostDetails matches the snapshot 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/testPostId"
  >
    Edit Post
  </Link>
  <Connect(Deleter)
    history={Object {}}
    id="testPostId"
    objectType="POST_TYPE"
  />
  <h2>
    Test Post Title
  </h2>
  <h3>
    Content: 
    testing PostDetails
  </h3>
  <h3>
    Author: 
    Test Post Author
  </h3>
  <h3>
    Time posted: 
    Wed Dec 31 1969 16:00:01 GMT-0800 (PST)
  </h3>
  <h3>
    Number of comments: 
    2
  </h3>
  <h3>
    Score: 
    <Connect(Voter)
      id="testPostId"
      objectType="POST_TYPE"
    />
  </h3>
  <h2>
    COMMENTS
  </h2>
  <Connect(Sorter) />
  <ul>
    <li>
      <Connect(Comment)
        category="testCategory"
        commentId="testComment1"
      />
    </li>
    <li>
      <Connect(Comment)
        category="testCategory"
        commentId="testComment2"
      />
    </li>
  </ul>
  <h4>
    <Link
      replace={false}
      to="/create/testCategory/testPostId"
    >
      New Comment
    </Link>
  </h4>
  <h4>
    <Link
      replace={false}
      to="/testCategory"
    >
      View 
      testCategory
       Category
    </Link>
  </h4>
  <h4>
    <Link
      replace={false}
      to="/"
    >
      View All Categories
    </Link>
  </h4>
</div>
`;
